### DISCUSSION TITLE
1주차 사전과제, 김경진

### DISCUSSION THREAD

**Discussion by lessmadmax** [MrMirror21/2024-KHU-CSE-TrackStudy-Web#17] [General]
**1. 명명 규칙(Naming Convention)이라는 것이 있습니다. 어떤 규칙들이 있는지, 각 규칙들은 어떻게 코드를 작성하는지, 자신이 그동안 사용해왔던 명명 규칙은 무엇인지 적어주세요.**

1. 카멜 케이스 : 첫 번째 단어는 소문자로 시작하고, 이후 각 단어의 첫 글자는 대문자로 표기한다.
2. 파스칼 케이스 :  모든 단어의 첫 글자를 대문자로 표기한다.
3. 스네이크 케이스 : 모든 글자를 소문자로 하고, 단어 사이를 언더스코어(_)로 구분한다.
4. 케밥 케이스 : 모든 글자를 소문자로 하고, 단어 사이를 하이픈(-)으로 구분한다.
5. 헝가리안 표기법 : 변수의 데이터 타입이나 목적을 접두어로 추가한다.

저는 이때까지 카멜 케이스를 주로 사용했습니다.

**2. 함수가 순수하다는 것은 무엇을 의미할까요?**
순수 함수란 다음 두 가지 조건을 만족하는 함수를 말한다. 

1. 참조 투명성 : 동일한 입력값에 대해 항상 동일한 출력값을 반환합니다. 즉, 함수의 결과가 입력에만 의존하며, 외부 상태나 변수에 의존하지 않습니다.
2. 부작용 없음 : 함수 내부에서 외부 상태를 변경하지 않습니다. 이는 함수가 외부 변수나 데이터 구조를 수정하지 않으며, 외부 입출력(IO) 작업(파일 쓰기, 네트워크 요청 등)을 수행하지 않음을 의미합니다.

**3. 개발 시에 순수 함수(pure function)로만 모든 코드를 작성하는 것에 대한 생각을 적어주세요.**

출력이 일관되고, 소프트웨어 공학적인 관점에서 결합도가 낮기 때문에 바람직하다고 생각합니다. 

**4. 자바스크립트에서 함수를 작성하는 문법은 다음과 같습니다. 이 외의 함수를 작성하는 문법 두 가지가 무엇이며, 어떤 때 사용하나요?**
```
function myFunc(param) {
	statements
}
```

1. 함수 표현식 : 변수에 함수를 할당하고, 이후에 함수를 사용할 때 해당 변수를 호출하는 방식으로 함수를 사용함. 해당 문법은 어떠한 함수가 다른 함수의 인자로 전달될 때 자주 사용되거나 함수가 정의되자마자 실행되도록 할 때 사용됨.
2. 화살표 함수 : 변수에 함수를 할당하는 것은 함수 표현식과 같으나, 
`const myFunc = 인수 => 인수 * 2;
`
와 같은 방식으로 쓴다. 해당 문법은 코드를 간결하게 하고 싶을 때 주로 사용한다.

**5. 파이썬에서는 문자열을 포맷팅하기 위해 다음과 같은 f-string을 사용합니다. 자바스크립트에서 f-string과 같이 문자열을 다양한 자료형과 변수들을 혼합하여 사용하려면 어떤 방법을 사용하나요?**

```
var x = 1;
var y = 2;
var result = `${x} + ${y}는 ${x + y}입니다.`;
console.log(result);  // '1 + 2는 3입니다.'
```

**6. 다음과 같은 객체(object)에서 p 값을 출력하려면 o.p 와 같은 방식으로 출력해야 하는데, p 로 출력할 수 있는 방법은 없을까요?**

```
var o = { p: 42, q: true };

console.log(o.p); // 42
console.log(p); // p is undefined
```
with 구문을 다음과 같이 사용하면 p로 출력할 수 있습니다. 
```
var o = { p: 42, q: true };

with (o) {
    console.log(p);
}
```

**Comment by MrMirror21** [comment]
답변해주신 사전 과제 잘 봤습니다.
1번에서 보통 카멜,파스칼,스네이크 케이스 정도가 유명한데 케밥 케이스와 헝가리안 표기법까지 자세하게 잘 조사해주셨네요. 👍
2번의 경우 답변해주신 내용중 side effect 용어를 부작용으로 번역하신 것 같으나, 실제로는 그냥 사이드 이펙트라고 많이 칭합니다.
6번에서 답변주신 with 문법도 정답이 될 수 있으나, 2009년 도입된 ES5 버전 이후의 자바스크립트에서는 변수 범위를 헷갈리게 하는 가독성 문제 때문에 strict mode에서 사용이 금지된 문법입니다. 해당 문제에 대한 답변을 스터디 중에 언급 드릴테니 잘 들어주시면 도움이 될 것 같습니다.
